### PR TITLE
[VIS] Working interactive resource graph using output json

### PR DESCRIPTION
**Pull request by augusta100**
# Summary
Resources graph with selectable teams and turns. Shows end of seasons.

## Additional Information
Need to change data source later after #230 merged

## Test Plan

- idk how to test stuff sorry :/

Screenshot 2021-01-02 at 18 22 54

### CODE REVIEW

**Comment by lhl2617** [comment]
we're not doing tests on vis so it's fine, if you're convinced it works well with a wide range of inputs we good.

File: website/tsconfig.json [hunk 1]
```diff
@@ -18,7 +18,8 @@
     "resolveJsonModule": true,
     "isolatedModules": true,
     "noEmit": true,
-    "jsx": "react-jsx"
+    "jsx": "react-jsx",
+    "noImplicitAny": false
```
Comment (lhl2617) [inline website/tsconfig.json:22]: hmmm, this isn't good, can you remove and solve the problems where you need implicit any's?

File: website/src/components/visualisation/Resources/Resources.tsx [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
+import LineRechartComponent from './LineGraph';
 
 const Resources = () => {
   return (
     <div className={styles.root}>
-      <img src={logo} className={styles.appLogo} alt="logo" />
-      <p className={styles.text}>
-        Self-Organising Multi-Agent Systems 2020
-        </p>
+      <h1>Resources over Time </h1>
```
Comment (lhl2617) [inline website/src/components/visualisation/Resources/Resources.tsx]:
```suggestion
      <h1>Resources over Time</h1>
```

File: website/src/components/visualisation/Resources/LineGraph.tsx [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            </ResponsiveContainer >
+        )
+    };
+}
+
+
+export default LineRechartComponent;
+
+
+
```
Comment (lhl2617) [inline website/src/components/visualisation/Resources/LineGraph.tsx]: if you're using vscode `cmd+shift+f` should autoformat and remove these lines

File: website/tsconfig.json (hunk 1 above)
Comment (augusta100) [reply website/tsconfig.json:22]: I'll have a look

File: website/src/components/visualisation/Resources/LineGraph.tsx [hunk 4]
```diff
@@ -0,0 +1,147 @@
+import React from "react";
+import { Brush, ResponsiveContainer, LineChart, Line, YAxis, XAxis, CartesianGrid, Tooltip, Legend, Surface, Symbols, ReferenceLine } from '../../../../node_modules/recharts';
```
Comment (lhl2617) [inline website/src/components/visualisation/Resources/LineGraph.tsx]:
```suggestion
import { Brush, ResponsiveContainer, LineChart, Line, YAxis, XAxis, CartesianGrid, Tooltip, Legend, Surface, Symbols, ReferenceLine } from 'recharts';
```
don't need to specify the direct path if it's a node module :)

File: website/src/components/visualisation/Resources/LineGraph.tsx [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+import outputJSON from '../../../output/output.json';
+import _ from "../../../../node_modules/lodash";
```
Comment (lhl2617) [inline website/src/components/visualisation/Resources/LineGraph.tsx]:
```suggestion
import _ from "lodash";
```

File: website/src/components/visualisation/Resources/LineGraph.tsx [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+
+class LineRechartComponent extends React.Component<{}, any> {
```
Comment (lhl2617) [inline website/src/components/visualisation/Resources/LineGraph.tsx]: haven't seen class components in some time... but this actually looks like a neat use case for class, I'm fine with this. We usually use hooks as per other files.

(I started react with class components, and still have a soft spot for them!)

File: website/src/components/visualisation/Resources/LineGraph.tsx [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        if(_.includes(this.state.disabled, dataKey)) {
+            this.setState({
+                disabled: this.state.disabled.filter(obj => obj !== dataKey)
+            });
+        }
+        else {
+            this.setState({
+                disabled:this.state.disabled.concat(dataKey)
+            });
+        }
```
Comment (lhl2617) [inline website/src/components/visualisation/Resources/LineGraph.tsx]: Don't think you need lodash here
```
// nice object decomposition to make it cleaner
const { disabled } = this state
if (disabled.includes(dataKey)) {
   this.setState( ... )
}
else {
   this.setState( ... )
}
```

File: website/src/components/visualisation/Resources/LineGraph.tsx [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    };
+
+    renderCustomizedLegend = ({ payload }) => {
+        return(
+            <div className="customized-legend">
+                {payload.map(entry => {
+                    const { dataKey, colour } = entry;
+                    const active = _.includes(this.state.disabled, dataKey);
```
Comment (lhl2617) [inline website/src/components/visualisation/Resources/LineGraph.tsx]: similarly, no lodash needed

File: website/src/components/visualisation/Resources/LineGraph.tsx [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        }
+        return seasonEnds;
+    };
+
+    render() {
+        return (
+            <ResponsiveContainer height={330} width = "95%">
+                <LineChart data={this.state.chartData.GameStates}
+                    margin={{ top: 5, right: 30, left: 20, bottom: 5 }}>
+                    {_.toPairs(this.state.lineColours)
```
Comment (lhl2617) [inline website/src/components/visualisation/Resources/LineGraph.tsx]: hmm -- does this look indented ok to you? also it looks to have a lot of logic, so we could make this a class fn

Comment (lhl2617) [reply website/src/components/visualisation/Resources/LineGraph.tsx]: @olly-larkin i think your diff can benefit from` _.toPairs`, seems neat

**Comment by lhl2617** [comment]
good job, overall looks ok to me.
will approve later after we merge in #230 - #232.

you also need to update from main

File: website/src/components/visualisation/Resources/LineGraph.tsx [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+class LineRechartComponent extends React.Component<{}, any> {
+
+    //data = outputJSON;
+
+    constructor(props){
+        super(props);
+        this.state = {
+            disabled: [],
+            lineColours: {
+                team1: "#0095FF",
```
Comment (lhl2617) [inline website/src/components/visualisation/Resources/LineGraph.tsx]: #234 merged, so we don't need to hardcode these now. for colours it's a bit of a pain but just put in some random colours after?

Comment (lhl2617) [reply website/src/components/visualisation/Resources/LineGraph.tsx]: (I'd personally write soemthing that generates colours based on a spectrum, but that's too much work. you can just make it loop around 6 colours as well if it's labelled ok.

File: website/src/components/visualisation/Resources/LineGraph.tsx (hunk 9 above)
Comment (ahm116) [reply website/src/components/visualisation/Resources/LineGraph.tsx:101]: Do you mean a class function for LineChart? Or a class function the Line lines in the chart?

Comment (ahm116) [reply website/src/components/visualisation/Resources/LineGraph.tsx:101]: And the tabbing in those four line 98-101 are right but the tabbing below was wrong. I did the autoformat  thing so should all be tabbed properly now

Comment (augusta100) [reply website/src/components/visualisation/Resources/LineGraph.tsx]: > hmm -- does this look indented ok to you? also it looks to have a lot of logic, so we could make this a class fn

The indenting is right for 98-101 but was wrong below that - I did the autoformat thing so should all be fine now. Do you mean a class fn for LineChart or just the Line lines?

Comment (lhl2617) [reply website/src/components/visualisation/Resources/LineGraph.tsx]: the lines from 101 onwards

File: website/tsconfig.json (hunk 1 above)
Comment (Darrekt) [reply website/tsconfig.json:22]: Feel free to ping us on discord if you need advice on how to type stuff, it's a bit tricky at first.

File: website/src/components/visualisation/Resources/LineGraph.tsx (hunk 3 above)
Comment (Darrekt) [reply website/src/components/visualisation/Resources/LineGraph.tsx]: @ahm116 might be `ctrl (command) + shift + i`, I think lh has a custom keybind.

Additionally, some autoformatters (prettier) think that rogue newlines are user-intended and leave them alone. It's good to check manually for this. In general for typescript, 1 line between every code block. New classes are usually in a separate file, so no need to do the "2 lines between classes, 1 between functions" thing that python does.

File: website/src/components/Visualisations/Resources/LineGraph.tsx [hunk 11]
```diff
@@ -0,0 +1,143 @@
+import React from "react";
+import { Brush, ResponsiveContainer, LineChart, Line, YAxis, XAxis, CartesianGrid, Tooltip, Legend, Surface, Symbols, ReferenceLine } from 'recharts';
```
Comment (Darrekt) [inline website/src/components/Visualisations/Resources/LineGraph.tsx]: Large imports like this usually get split into new lines, or get namespaced. Are you using Prettier?

File: website/src/components/Visualisations/Resources/LineGraph.tsx [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        }
+    };
+
+    renderCustomizedLegend = ({ payload }) => {
+        return (
+            <div className="customized-legend">
+                {payload.map(entry => {
+                    const { dataKey, colour } = entry;
+                    const { disabled } = this.state;
+                    const active = disabled.includes(dataKey);
```
Comment (Darrekt) [inline website/src/components/Visualisations/Resources/LineGraph.tsx]:
```suggestion
                    const active = this.state.disabled.includes(dataKey);
```

Comment (Darrekt) [reply website/src/components/Visualisations/Resources/LineGraph.tsx]: Doesn't seem like you use `disabled` anywhere else

File: website/src/components/Visualisations/Resources/LineGraph.tsx [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+                team5: "ClientInfos.Team5.Resources",
+                team6: "ClientInfos.Team6.Resources",
+            },
+            chartData: outputJSON
+        };
+    }
+
+    handleClick = dataKey => {
+        const { disabled } = this.state;
+        if (disabled.includes(dataKey)) {
```
Comment (Darrekt) [inline website/src/components/Visualisations/Resources/LineGraph.tsx:35]:
```suggestion
        if (this.state.disabled.includes(dataKey)) {
```

Comment (Darrekt) [reply website/src/components/Visualisations/Resources/LineGraph.tsx:35]: In fact, you still use `this.state.disabled` below.

File: website/src/components/Visualisations/Resources/LineGraph.tsx [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+                    )
+                })}
+            </div>
+        );
+    };
+
+    getSeasonEnds() {
+        var seasonEnds: any[] = [];
+        let i;
+        for (i in this.state.chartData.GameStates) {
```
Comment (Darrekt) [inline website/src/components/Visualisations/Resources/LineGraph.tsx:100]: We have `forEach` in javascript on arrays, give it a try!

**Review by Darrekt — CHANGES_REQUESTED** [review]
Looks good, I just have a bunch of pedantic remarks. See comments.

You haven't hooked this up to the actual navigator, would love to see that and try running it on my local branch. (we can help you with this, just let us know).

File: website/src/components/Visualisations/Resources/LineGraph.tsx [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        return (
+            <div className="customized-legend">
+                {payload.map(entry => {
+                    const { dataKey, colour } = entry;
+                    const { disabled } = this.state;
+                    const active = disabled.includes(dataKey);
+                    const style = {
+                        marginRight: 10,
+                        colour: active ? "#AAA" : "#000"
+                    };
```
Comment (Darrekt) [inline website/src/components/Visualisations/Resources/LineGraph.tsx]: Not important, but would like to discuss for fun:

I usually like declaring constants like this at the top, so in this case between l47 & l48. I feel it makes code a bit clearer that the constants aren't part of the imperative looping logic.

File: website/src/components/Visualisations/Resources/LineGraph.tsx [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+        if (disabled.includes(dataKey)) {
+            this.setState({
+                disabled: this.state.disabled.filter(obj => obj !== dataKey)
+            });
+        }
+        else {
+            this.setState({
+                disabled: this.state.disabled.concat(dataKey)
+            });
+        }
```
Comment (Darrekt) [inline website/src/components/Visualisations/Resources/LineGraph.tsx]:
```suggestion
this.setState({
      disabled: this.state.disabled.includes(dataKey)
        ? this.state.disabled.filter((obj) => obj !== dataKey)
        : this.state.disabled.concat(dataKey),
    });
```

My indentation is broken but you get the point

**Review by augusta100 — COMMENTED** [review]
Added all these changes

**Review by Darrekt — APPROVED** [review]
S T A M P

**Pull request merged**